src/fluree/db/query/exec/where.cljc

```diff
@@ -0,0 +1,338 @@
+
+(defn with-distinct-subjects
+  "Return a transducer that filters a stream of flakes by removing any flakes with
+  subject ids repeated from previously processed flakes."
+  []
+  (fn [rf]
+    (let [seen-sids (volatile! #{})]
+      (fn
+        ;; Initialization: do nothing but initialize the supplied reducing fn
+        ([]
+         (rf))
+
+        ;; Iteration: keep track of subject ids seen; only pass flakes with new
+        ;; subject ids through to the supplied reducing fn.
+        ([result f]
+         (let [sid (flake/s f)]
+           (if (contains? @seen-sids sid)
+             result
+             (do (vswap! seen-sids conj sid)
+                 (rf result f)))))
+
+        ;; Termination: do nothing but terminate the supplied reducing fn
+        ([result]
+         (rf result))))))
+
+(defmethod match-pattern :class
+  [db solution pattern filters error-ch]
+  (let [triple   (val pattern)
+        [s p o]  (assign-matched-values triple solution filters)
+        cls      (::val o)
+        classes  (into [cls] (dbproto/-class-prop db :subclasses cls))
+        class-ch (async/to-chan! classes)
+        match-ch (async/chan 2 (comp cat
+                                     (with-distinct-subjects)
+                                     (map (fn [flake]
+                                            (match-flake solution triple flake)))))]
+    (async/pipeline-async 2
+                          match-ch
+                          (fn [cls ch]
+                            (-> (resolve-flake-range db error-ch [s p (assoc o ::val cls)])
+                                (async/pipe ch)))
+                          class-ch)
+    match-ch))
+
+(defn with-constraint
+  "Return a channel of all solutions from `db` that extend from the solutions in
+  `solution-ch` and also match the where-clause pattern `pattern`."
+  [db pattern filters error-ch solution-ch]
+  (let [out-ch (async/chan 2)]
+    (async/pipeline-async 2
+                          out-ch
+                          (fn [solution ch]
+                            (-> (match-pattern db solution pattern filters error-ch)
+                                (async/pipe ch)))
+                          solution-ch)
+    out-ch))
+
+(defn match-clause
+  "Returns a channel that will eventually contain all match solutions in `db`
+  extending from `solution` that also match all the patterns in the parsed where
+  clause collection `clause`."
+  [db solution clause error-ch]
+  (let [initial-ch (async/to-chan! [solution])
+        filters    (::filters clause)
+        patterns   (::patterns clause)]
+    (reduce (fn [solution-ch pattern]
+              (with-constraint db pattern filters error-ch solution-ch))
+            initial-ch patterns)))
+
+(defmethod match-pattern :union
+  [db solution pattern _ error-ch]
+  (let [clauses   (val pattern)
+        clause-ch (async/to-chan! clauses)
+        out-ch    (async/chan 2)]
+    (async/pipeline-async 2
```

Comment:
I really love the use of `pipeline-async` wherever possible, it really makes it easier to reason about what's happening